### PR TITLE
[MLIR,Python] Support converting boolean numpy arrays to and from mlir attributes

### PR DESCRIPTION
**Pull request by kasper0406**
Currently it is unsupported to:
1. Convert a `MlirAttribute` with type `i1` to a numpy array
2. Convert a boolean numpy array to a `MlirAttribute`

Currently the entire Python application violently crashes with a quite poor error message https://github.com/pybind/pybind11/issues/3336

The complication handling these conversions, is that `MlirAttribute` represent booleans as a bit-packed `i1` type, whereas numpy represents booleans as a byte array with 8 bit used per boolean.

This PR proposes the following approach:
1. When converting a `i1` typed `MlirAttribute` to a numpy array, we can not directly use the underlying raw data backing the `MlirAttribute` as a buffer to Python, as done for other types. Instead, a copy of the data is generated using numpy's unpackbits function, and the result is send back to Python.
2. When constructing a `MlirAttribute` from a numpy array, first the python data is read as a `uint8_t` to get it converted to the endianess used internally in mlir. Then the booleans are bitpacked using numpy's bitpack function, and the bitpacked array is saved as the `MlirAttribute` representation.

Please note that I am not sure if this approach is the desired solution. I'd appreciate any feedback.

### CODE REVIEW

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (kasper0406) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:762]: I had to move the if-statements to the `getAttributeFromBuffer` method, as now the i1 case will not follow the usual flow, but instead call `getBitpackedAttributeFromBooleanBuffer` to construct the MlirAttribute.

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1041]:
```suggestion
  // There is a complication for boolean numpy arrays, as numpy represents them
```

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1042]:
```suggestion
  // as 8 bits (1 byte) per boolean, whereas MLIR bitpacks them into 8 booleans per byte.
```

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:981]: sorry dumb question: why doesn't `isSignedIntegerFormat(format)` currently work for `i1`?

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1048]: sorry don't understand? what/why does this need to be "corrected"? I can see you're setting `little` below but shouldn't you be preserving/using the endianness for the host arch (wherever the bindings are running...)? probably I don't understand something here...

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1094]: why this change?

File: mlir/lib/Bindings/Python/PybindUtils.h
Comment (makslevental) [inline mlir/lib/Bindings/Python/PybindUtils.h:17]: unused i think?

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:981]: I believe it is because of the bit-packing. If it started returning true, I think that it will not be possible to distinguish between the i8 (and u8) cases and the boolean case, as they both would have an itemsize of 1.
So the format would have to be compared to "?" explicitly anyways.

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1048]: From https://mlir.llvm.org/docs/BytecodeFormat/#fixed-width-integers my understanding is that MLIR attributes always have their data stored in a little-endian format. If that is not true for MlirAttributes in general, the code is wrong.

File: mlir/lib/Bindings/Python/PybindUtils.h
Comment (kasper0406) [reply mlir/lib/Bindings/Python/PybindUtils.h:17]: It is used in `IRAttributes.cpp` now to use the numpy bitpacking functions. I didn't include it directly in `IRAttributes.cpp`, because it seems like all `pybind`` includes have been collected in `PybindUtils.h`

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp]: Typo

File: mlir/lib/Bindings/Python/PybindUtils.h
Comment (makslevental) [reply mlir/lib/Bindings/Python/PybindUtils.h:17]: nah definitely (as a rule) put `include`s in cpp (ie implementation) unless it's absolutely necessary to put them in the headers. Even then, we often "pre declare" instead of including the header.

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1060]: nit: camel case instead of snake

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1062]: nit: camelCase instead of snake_case

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1061]: ................ you can do this `"bitorder"_a = "little"` - wow call me shocked I never noticed/knew you could do "kwargs" like this on cpp side.

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [inline mlir/lib/Bindings/Python/IRAttributes.cpp:1084]: same nit from above

**Review by makslevental — COMMENTED** [review]
Looks basically fine/good to me (modulo nits). @stellaraccident do you have any complaints/concerns here?

File: mlir/lib/Bindings/Python/IRAttributes.cpp
Comment (makslevental) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1048]: @stellaraccident this is the one i'm most uncertain about

Comment (jpienaar) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: That's for bytecode rather than in memory. If this is converting the in memory format, then that section doesn't apply.

Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: Ack, one thing I don't understand though, is that if I follow the `mlirDenseElementsAttrRawBufferGet` function, it goes to `DenseElementsAttr::getFromRawBuffer`, and I believe it eventually ends up in `BuiltinAttributes.cpp` where it uses the [`writeBits` and `readBits` functions](https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/BuiltinAttributes.cpp#L541) respectively. Both of these have special handling for big endian, where it seems to convert the format.

It is possible that it ends up [calling `Base::get` instead](https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/BuiltinAttributes.cpp#L1354), which may be why the above is not true. My C++ navigation skills is currently failing me, to follow which code this `Base::get` call will execute.

Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: Do you know of a way to test this? From reading around other llvm docs, it seems like user-mode qemu is the common approach for testing big-endian systems. Do any of you have experience with that and the Python API?

Comment (makslevental) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: > it goes to `DenseElementsAttr::getFromRawBuffer`

https://github.com/makslevental/llvm-project/blob/84aa02d3fa1f1f614c4f3c144ec118b2f05ae6b0/mlir/lib/IR/BuiltinAttributes.cpp#L1064

which does indeed hit `Base::get`

https://github.com/makslevental/llvm-project/blob/84aa02d3fa1f1f614c4f3c144ec118b2f05ae6b0/mlir/lib/IR/BuiltinAttributes.cpp#L1354

which then goes to `static ConcreteT StorageUserBase::get`

https://github.com/makslevental/llvm-project/blob/a9636b7f60f283926c66e96c036f5b5d9e57c026/mlir/include/mlir/IR/StorageUniquerSupport.h#L177

I think possibly you could just remove the explicit use of `little` in the various places and defer to host endiannes? I mean really this is all academic because as I understand it, there are no extant big-endian systems we care about? Someone with more experience can comment on this hypothesis...

Comment (kasper0406) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: Thanks for the pointers!

Unfortunately, it will not work simply remove the "little" part from the numpy calls, as it defaults to big-endian and not the native byteorder.

As you mention, I doubt any big-endian users will ever call this code anyways. What do you think about just throwing a "this is unsupported" exception (which is the current behaviour) if someone calls it on a big-endian system?

I think that check can be made something like:
```c++
if (llvm::endianness::native == llvm::endianness::big) {
    throw py::type_error("Boolean types are unsupported on big-endian systems");
}
```

I could also use the above check to switch between "little"/"big" endian in the numpy call, but I am considering if throwing is actually better, given we have no way to really test it?

Comment (makslevental) [reply mlir/lib/Bindings/Python/IRAttributes.cpp:1049]: Ya throwing here is acceptable and reasonable (if why IBM customers ever show up we can discuss it then 😅)

**Review by makslevental — APPROVED** [review]
LGTM - sorry for the long arduous process. I'll try to be faster next time

**Comment by kasper0406** [comment]
Awesome, thank you for the review and comments!
I'd appreciate if you can click the merge button, as I don't have the permissions to do so.

**Comment by makslevental** [comment]
> LLVM Buildbot has detected a new failure on builder `sanitizer-x86_64-linux-bootstrap-asan` running on `sanitizer-buildbot1` while building `mlir` at step 2 "annotate".


Flake I'm pretty sure

**Comment by makslevental** [comment]
> then extracts a pointer into it:
> 
> ```c++
>     py::buffer_info pythonBuffer =
>         unpacked_booleans.cast<py::buffer>().request();
> ```
> 

Ah I believe there's some API equivalent to `release()` because we do munge `py:buffer`s elsewhere...

> I'm going to revert the patch now.

Sure - I can stamp.

**Comment by gribozavr** [comment]
I pushed the revert as https://github.com/llvm/llvm-project/commit/0a68171b3c67503f7143856580f1b22a93ef566e.

**Comment by makslevental** [comment]
here's the dance you have to do to release https://github.com/llvm/llvm-project/blob/6b64f365364a6bf9a0ae039f6115871bc3b8ce07/mlir/lib/Bindings/Python/IRAttributes.cpp#L1228-L1239 but you will also have to think about https://pybind11.readthedocs.io/en/stable/advanced/functions.html#return-value-policies

I'll try to reason it out ~tomorrow

**Comment by kasper0406** [comment]
Agh dang. I will try to give it a look tomorrow.
Are there any information available for how to get the ASan report? I tried building with `-DLLVM_USE_SANITIZER=Address`, but I do not see any issues running `ninja -C build check-mlir` with that option.

**Comment by joker-eph** [comment]
> I'm afraid there is an actual ASan issue here. While the above one is a flake, I am getting an ASan failure that is real.

@gribozavr : is this an upstream test failing or something downstream here? If downstream can you provide a reproducer please?

**Pull request merged**
